File: 2023/UICPC Regional/Round #9/solutions/numbers/submissions/accepted/numbers_zf.py
#!/usr/bin/python3

def solve(str):
    if len(str) == 1 or str == "22": return str

    # lazy thinking, can probably combine the two loops into 1
    i = 0
    mod4 = 0
    while i < len(str) or mod4:
        if (i == len(str) or str[i] != '2') and mod4:
            str = str[:i] + '2' + str[i:]
            mod4 = 0
        else:
            mod4 = (mod4 + int(str[i]))%4;
        i = i+1

    i = 0
    mod8 = 0
    while i < len(str) or mod8:
        if (i == len(str) or str[i] == '8') and mod8:
            str = str[:i] + '4' + str[i:]
            mod8 = 0
        else:
            mod8 = (mod8 + int(str[i]))%8
        i = i+1

    tot = 0
    for c in str:
        tot += int(c)

    while tot&(tot-1):
        str += "8"
        tot += 8

    return str

t = int(input())
for i in range(t):
    str = input()
    print(solve(str))

File: 2023/UICPC Regional/Round #9/solutions/numbers/submissions/accepted/numbers_zf.cpp
#include <iostream>
#include <string>

using namespace std;

string solve() {
    string s;
    cin >> s;

    if (s.length() == 1 || s == "22") return s;


    for (int i = 0, cnt = 0; i <= s.length(); ++i) {
        if ((i == s.length() || s[i] > '2') && cnt&2) s.insert(i, "2");
        else if ((i == s.length() || s[i] > '4') && cnt&4) s.insert(i, "4");
        cnt += s[i]-'0';
    }

    int tot = 0;
    for (int i = 0; i < s.length(); ++i) tot += s[i]-'0';
    while (tot&(tot-1)) {
        tot += 8;
        s += "8";
    }

    return s;
}

int main() {
    int t;
    cin >> t;
    while (t--) {
        cout << solve() << endl;
    }

    return 0;
}
